forte/solvers/solver.py

```diff
@@ -1,31 +1,117 @@
+    or provided by a solver.
+    """
+    MODEL = auto()
+    ORBITALS = auto()
+    RDMS = auto()
+
+
 class Solver(ABC):
     """
-    A class used to implement a quantum chemistry solver.
+    Represents a node on a computational graph that performs
+    an operation.
+
+    A Solver is a node in a graph that takes zero or more input
+    solvers. This class is used to implement quantum chemistry
+    methods and other operations.
+    This class implements a ``run()`` method that can be called
+    to start the evaluation of the computational graph.
+
+    To check if the validity of a graph, each node provides a
+    list of features that are needed and provided.
+    The features that are needed must be part of the input node(s).
 
     Solver stores Forte base objects in a data attribute
     and a results object.
     """
-    def __init__(self):
+    def __init__(self, needs, provides, input=None, options=None, cbh=None):
+        """
+        Parameters
+        ----------
+        needs: list(str)
+            a list of features required by this solver
+        provides: list(str)
+            a list of features provided by this solver
+        input: list(Solver)
+            a list of input nodes to this node
+        options: dict(str -> obj)
+            a dictionary of options to pass to the forte modules
+        cbh: CallbackHandler
+            a callback handler object
+        """
+        self._needs = needs
+        self._provides = provides
+        # input can be None, a single element, or a list
+        input = [] if input is None else input
+        self._input = [input] if type(input) is not list else input
+        self._check_input()
+
+        self._options = {} if options is None else options
+        self._cbh = CallbackHandler() if cbh is None else cbh
         self._executed = False
         self._data = Data()
         self._results = Results()
+        # the default psi4 output file
         self._output_file = 'output.dat'
+        # self._output_file = f'output.{time.strftime("%Y-%m-%d-%H:%M:%S")}.dat'
 
-    @abstractmethod
+    # decorate to increase the log depth
+    @increase_log_depth
     def run(self):
-        """Interface for running the solver."""
+        """
+        A general solver interface.
+
+        This method is common to all solvers, and in turn it is routed to
+        the method ``_run()`` implemented differently in each solver.
+
+        Return
+        ------
+            A data object
+        """
+        # log call to run()
+        flog('info', f'{type(self).__name__}: calling run()')
+
+        # call derived class implementation of _run()
+        self._run()
+
+        # log end of run()
+        flog('info', f'{type(self).__name__}: run() finished executing')
+
+        # set executed flag
+        self._executed = True
+
+        return self.data
+
+    @abstractmethod
+    def _run():
+        """The actual run function implemented by each method"""
```

Comment:
```suggestion
        """The actual run function implemented by each method"""
        pass
```

Reply:
I'll implement the `pass` but from what I know the minimum text needed for a function to be fine is either a docstring or `pass`, and both are equivalent in the sense that they don't perform any operation.